pkg/placement/set-types.go

```diff
@@ -0,0 +1,220 @@
+		}
+		return nil
+	})
+	return
+}
+
+func SetRemoveAll[Elt comparable](set MutableSet[Elt], goners Visitable[Elt]) (someNew, allNew bool) {
+	someNew, allNew = false, true
+	goners.Visit(func(goner Elt) error {
+		if set.Remove(goner) {
+			someNew = true
+		} else {
+			allNew = false
+		}
+		return nil
+	})
+	return
+}
+
+// SetChangeReceiverReverse returns a receiver that acts in the opposite way as the given receiver.
+// That is, Add and Remove are swapped.
+func SetChangeReceiverReverse[Elt comparable](forward SetChangeReceiver[Elt]) SetChangeReceiver[Elt] {
+	return setChangeReceiverReverse[Elt]{forward}
+}
+
+type setChangeReceiverReverse[Elt comparable] struct {
+	forward SetChangeReceiver[Elt]
+}
+
+// SetChangeReceiverFork constructs a SetChangeReceiver that broadcasts incoming changes to the given receivers.
+// Each change is passed on to every receiver, and
+// combineWithAnd says whether to combine the returned values with AND
+// (the alternative is OR).
+func SetChangeReceiverFork[Elt comparable](combineWithAnd bool, receivers ...SetChangeReceiver[Elt]) SetChangeReceiver[Elt] {
+	return &setChangeReceiverFork[Elt]{
+		combineWithAnd: combineWithAnd,
+		receivers:      receivers,
+	}
+}
+
+type setChangeReceiverFork[Elt comparable] struct {
+	combineWithAnd bool
+	receivers      []SetChangeReceiver[Elt]
+}
+
+func (crf *setChangeReceiverFork[Elt]) Add(elt Elt) bool {
+	ans := crf.combineWithAnd
+	for _, receiver := range crf.receivers {
+		change := receiver.Add(elt)
+		if crf.combineWithAnd {
+			ans = ans && change
+		} else {
+			ans = ans || change
+		}
+	}
+	return ans
+}
+
+func (crf *setChangeReceiverFork[Elt]) Remove(elt Elt) bool {
+	ans := crf.combineWithAnd
+	for _, receiver := range crf.receivers {
+		change := receiver.Remove(elt)
+		if crf.combineWithAnd {
+			ans = ans && change
+		} else {
+			ans = ans || change
+		}
+	}
+	return ans
+}
+
+func (crr setChangeReceiverReverse[Elt]) Add(elt Elt) bool {
+	return crr.forward.Remove(elt)
+}
+
+func (crr setChangeReceiverReverse[Elt]) Remove(elt Elt) bool {
+	return crr.forward.Add(elt)
+}
+
+// Reducer is something that crunches a collection down into one value
+type Reducer[Elt any, Ans any] func(Visitable[Elt]) Ans
+
+// NewReducer makes Reducer that works with accumulator values
+func ValueReducer[Elt any, Accum any, Ans any](initialize func() Accum, add func(Accum, Elt) Accum, finish func(Accum) Ans) Reducer[Elt, Ans] {
+	return func(elts Visitable[Elt]) Ans {
+		accum := initialize()
+		elts.Visit(func(elt Elt) error {
+			accum = add(accum, elt)
+			return nil
+		})
+		return finish(accum)
+	}
+}
+
+// Identity1 is useful in reduers where the accumulator has the same type as the result
+func Identity1[Val any](val Val) Val { return val }
+
+// StatefulReducer makes a Reducer that works with a stateful accumulator
+func StatefulReducer[Elt any, Accum any, Ans any](initialize func() Accum, add func(Accum, Elt), finish func(Accum) Ans) Reducer[Elt, Ans] {
```

Comment:
Looks like the only difference between `ValueReducer` and `StatefulReducer` is the signature of `add`. May I have more info in what sense that difference makes it 'stateful' here?

Reply:
Thanks!